File: food-truck-api/src/main/java/food/truck/api/reviews_and_subscriptions/Review.java
package food.truck.api.reviews_and_subscriptions;


import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import food.truck.api.truck.Truck;
import lombok.Data;
import org.jetbrains.annotations.Nullable;

import javax.persistence.*;
import java.sql.Timestamp;

@Data
@Entity
@Table(name="review")
public class Review {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "review_id", nullable = false)
    Long id;

    @JoinColumn(foreignKey = @ForeignKey(name = "user_id"), nullable = false)
    Long userId;

    @ManyToOne
    @JoinColumn(name = "truck_id", nullable = false)
    @JsonIgnore
    Truck truck;

    @JsonProperty("truckId")
    long truckId() {
        return truck.getId();
    }

    @Column(name = "star_rating", nullable = false)
    Integer starRating;

    @Column(name = "cost_rating", nullable = false)
    Integer costRating;

    @Column(name = "review_text")
    @Nullable
    String reviewText;

    @Column(name = "date", nullable = false)
    Timestamp day_time;


}

File: food-truck-frontend/src/domain/truck/Review.ts
interface Review {
  // Administrative stuff
  reviewId: number;
  userId: number;
  truckId: number;
  extended: boolean | undefined;

  // Information of interest to users
  username: string;
  starRating: number;
  costRating: number;
  review: string;
  timestamp: Date;
}

export function emptyReview(userId: number, truckId: number): Review {
  return {
    reviewId: -1,
    userId,
    truckId,
    extended: false,
    username: "",
    starRating: 0,
    costRating: 0,
    review: "",
    timestamp: new Date(),
  };
}

export function backendToFrontend(obj: any, username: string): Review {
  return {
    reviewId: obj.id,
    userId: obj.userId,
    truckId: obj.truck.id,
    extended: obj.reviewText? obj.reviewText.length > 0 : false,
    starRating: obj.starRating,
    costRating: obj.costRating,
    review: obj.reviewText,
    timestamp: new Date(obj.day_time),
    username,
  };
}

export function frontendToBackend(review: Review): any {
  return {
    reviewId: review.reviewId,
    userId: review.userId,
    truckId: review.truckId,
    starRating: review.starRating,
    costRating: review.costRating,
    review: (review.extended? review.review : ""),
    timestamp: review.timestamp,
  };
}

export default Review;
